lib/collection/src/collection_manager/optimizers/config_mismatch_optimizer.rs

```diff
@@ -116,13 +116,41 @@ impl ConfigMismatchOptimizer {
                         .any(|(vector_name, vector_data)| {
                             // Check HNSW mismatch
                             match &vector_data.index {
-                                Indexes::Plain {} => false,
+                                Indexes::Plain {} => {}
                                 Indexes::Hnsw(effective_hnsw) => {
                                     // Select segment if we have an HNSW mismatch that requires rebuild
                                     let target_hnsw = self.get_required_hnsw_config(vector_name);
-                                    effective_hnsw.mismatch_requires_rebuild(&target_hnsw)
+                                    if effective_hnsw.mismatch_requires_rebuild(&target_hnsw) {
+                                        return true;
+                                    }
                                 }
                             }
+
+                            // Check quantization mismatch
+                            let target_quantization_collection = self.quantization_config.as_ref();
+                            let target_quantization_vector = self
+                                .collection_params
+                                .vectors
+                                .get_params(vector_name)
+                                .and_then(|vector_params| {
+                                    vector_params.quantization_config.clone()
+                                });
+                            let target_quantization = target_quantization_vector
+                                .as_ref()
+                                .or(target_quantization_collection);
+                            let quantization_mismatch = vector_data
+                                .quantization_config
+                                .as_ref()
+                                .zip(target_quantization)
+                                // Rebuild if current parameters differ from new parameters
+                                .map(|(a, b)| a.mismatch_requires_rebuild(b))
+                                // Or rebuild if we currently have quantization but now disabled it
+                                .unwrap_or_else(|| {
+                                    vector_data.quantization_config.is_some()
```

Comment:
There are no conditions for enabling quantization, only for disabling.
I propose to change condition for both enabling and disabling:
`vector_data.quantization_config.is_some() != target_quantization.is_some()`

Reply:
I also tried to enable quantization locally from rest api, I failed because of this condition

Reply:
Thanks for giving it a shot :smiley: 

Changed it and testing locally now.

Reply:
Maybe my proposal fix is incorrect, need more investigation

Reply:
I have an idea why this happens. Plain segment has always empty quantization config, maybe it produces infinity loop (and that's why I have 0MB in logs - it's reloading of empty plain segment)

Reply:
What about this?

```rust
                                // Or rebuild if we now change the enabled state on an indexed segment
                                .unwrap_or_else(|| {
                                    vector_data.index.is_indexed()
                                        && (vector_data.quantization_config.is_some()
                                            != target_quantization.is_some())
                                });
```

Reply:
> What about this?
> 
> ```rust
> // Or rebuild if we now change the enabled state on an indexed segment
> .unwrap_or_else(|| {
>     vector_data.index.is_indexed()
>         && (vector_data.quantization_config.is_some() != target_quantization.is_some())
> });
> ```

I guess it's fine if it tested. Maybe we can add some separate condition for checking, is segment "quantizable" or not. But I'm not sure what option is better

Reply:
> Do we call this optimizer check or not?

Then we don't. Only if details change, in which case the optimizers are recreated.

Reply:
In this case, looks fine. Can you pls test it locally? I'll test it too with different configs

Reply:
Testing it now. Already did the unit and integration tests, those succeed. Trying some manual variants now.

Reply:
Results look positive: 
![image](https://github.com/qdrant/qdrant/assets/856222/20645966-ca8a-43fe-89f5-5131862f2962)

Reply:
Tested it locally and I'm seeing good results. I've pushed the change onto this branch.

@IvanPleshkov If your testing succeeds as well, could you resolve this comment and update the review accordingly?

Reply:
Checked locally, seems to be perfect! Resolve